### ISSUE TITLE
CoqIDE

### ISSUE THREAD

**Issue by QinxiangCao** [coq/coq#11843]

**Comment by Zimmi48** [comment]
Could not reproduce on Linux, with filename `test.v`.

**Comment by QinxiangCao** [comment]
I did more tests.

1. MacOS, CoqIDE on mac: no problem.
2. 64-bit windows. 32-bit/64-bit CoqIDE on windows, a short file name like "TestGenerateAfterCompile.v": only ".glob" is generated when pressing "compile". When closing CoqIDE, a new ".glob" and three extra files ".vo", ".vok", ".vos" are created.
3. 64-bit windows. 32-bit/64-bit CoqIDE on windows, a short file name like "TestNotCompileNameShouldBeLong.v": only ".glob" is generated when pressing "compile". When closing CoqIDE, a new ".glob" is created. ".vo" cannot be generated.
4. 64-bit windows. 32-bit/64-bit CoqIDE‘s bin/coqc on windows-Cygwin command line: no problem.


Also, if I use the following file (changed a little bit), the result of test2 above is like test3, no vo file can be generated.


Module Task1.

Hypothesis foo0: True.
Hypothesis foo1: True.
Hypothesis foo2: True.
Hypothesis foo3: True.
Hypothesis foo4: True.
Hypothesis foo5: True.
Hypothesis foo6: True.
Hypothesis foo7: True.
Hypothesis foo8: True.
Hypothesis foo9: True.

End Task1.



Module Task3.

Hypothesis foo0: True.
Hypothesis foo1: True.
Hypothesis foo2: True.
Hypothesis foo3: True.
Hypothesis foo4: True.
Hypothesis foo5: True.
Hypothesis foo6: True.
Hypothesis foo7: True.

End Task3.

**Comment by Zimmi48** [comment]
Thanks @QinxiangCao for the extensive testing. This is a very strange bug! @MSoegtropIMC any idea what could be going on?

**Comment by ppedrot** [comment]
The "Compile Buffer" command is known to be seriously buggy, see e.g. #3991. Can you try to run CoqIDE with the `-debug` flag and report additional output?

**Comment by QinxiangCao** [comment]
> The "Compile Buffer" command is known to be seriously buggy, see e.g. #3991. Can you try to run CoqIDE with the `-debug` flag and report additional output?

What does that mean? I tried "Edit -> Preferences -> externals" and changed "coqc" to "coqc -bebug". Is that what I am supposed to do? I don't see anything different in the message windows.

**Comment by ppedrot** [comment]
You need to run CoqIDE from a console and pass it the `-debug` flag there. I don't know how to do that properly on Windows though.

**Comment by QinxiangCao** [comment]
> You need to run CoqIDE from a console and pass it the `-debug` flag there. I don't know how to do that properly on Windows though.

Thank you. I can do that through windows cmd (but not cygwin command line). 

The log is attached.

[coqide_ffca2a.log](https://github.com/coq/coq/files/4349918/coqide_ffca2a.log)

**Comment by QinxiangCao** [comment]
Here is another one.
[coqide_1432fc.log](https://github.com/coq/coq/files/4350168/coqide_1432fc.log)

**Comment by ppedrot** [comment]
Unfortunately, it seems that the call to coqc is not part of the debug info. Maybe the error `Unix.Unix_error(40, "unix_set_nonblock", "")` is related to it, though, I'll have a look at the code.

**Comment by ppedrot** [comment]
Indeed, running the coqc command triggers a `set_nonblock`. I have no idea why this fails on Windows though.

**Comment by MSoegtropIMC** [comment]
I compared this with SysInternals ProcMon. There are no obvious differences, except that when called via CoqIDE, the coq bin folder is in the PATH and the working directory is different. But the execution is almost identical until it comes to reading the .v file. The process started via console reads it, does an feof test and closes it. The coqide process reads it but does not do a feof test before closing it. I attached a diff between the ProcMon logs (this is a HTML file, so change the extension and open in a browser).

What might happen is that nothing is connected to the stdout pipe and it depends on how much is written into the pipe if it fails or not. If the pipe buffer get's full (likely 4k - the message size I get with the extended test and my fiel path is 6k) with nothing connected to the other end it is game over. How much is written e.g. depends on the length of the file name and the number of warnings written. The warnings output by coqc look like:
```
File "D:\CoqInstallerTests\Issue_11843\test1.v", line 25, characters 0-22:
Warning: Interpreting this declaration as if a global declaration prefixed by
"Local", i.e. as a global declaration which shall not be available without
qualification when imported. [local-declaration,scope]
```
I didn't look at the code - is it the same for Linux and WIndows? If so it is likely an issue with the OCaml system command. As I said I have an alternative and could create a PR but in the past people found this too agressive a change (to replace the system command of OCaml on Windows with our own version, of couse I would try to get it upstream as well). If our code is different, I think we messed up I/O redirection on Windows.

[Report.txt](https://github.com/coq/coq/files/4350838/Report.txt)

**Comment by QinxiangCao** [comment]
That's right. This code will generate many warning messages. And in my tests, whether a silent compilation failure is triggered does depend on the number warnings, the length of files, the lengths of PATH, etc.

**Comment by MSoegtropIMC** [comment]
@QinxiangCao : could you please verify that the limit where it fails is 4k (capture the output of coqc)?

Since this seems to be Windows specific, I take this.

Any thoughts on the small fix vs. big fix question?

**Comment by QinxiangCao** [comment]
I think the answer is yes. I modify the file name so that the warning message is 4088 characters or 4133 characters. And CoqIDE compiles correctly for the former one, but fails on the latter one.

**Comment by Zimmi48** [comment]
@MSoegtropIMC Do you still intend to address this for 8.12? (This can also be done after the branching of course.)

**Comment by MSoegtropIMC** [comment]
Yes, I am just busy with the platform and reals stuff, but it should go into 8.12.

**Comment by Zimmi48** [comment]
This will have to be delayed to 8.12.0.

**Comment by Zimmi48** [comment]
@MSoegtropIMC friendly ping to remind you about this issue :wink:

**Comment by SkySkimmer** [comment]
@MSoegtropIMC ping

**Comment by MSoegtropIMC** [comment]
We already did quite a bit of root cause analysis - the issue seems to be a buffer limitation of 4K for warnings. Can someone else take it from here? I have little bandwidth left besides Coq Platform.
